### ISSUE TITLE
no motion trans positions in results.

### ISSUE THREAD

**Issue by XH-B** [setarehc/diffusion-motion-inbetweening#11]
When I was running this script `(python-m sample-conditional-synthesis-model-path./save/endmdi-urandomframes/model000750000. pt -- edit_mode uncond -- num_Samples 10-- num_deditions 3`), the generated result did not have tans, only bone rotation. How was your visualization result generated。Thanks

**Comment by setarehc** [comment]
Can you elaborate on exactly what you mean by output? Do you mean the raw sample which is output from the diffusion model or the processed sample that is used for visualization? The raw sample has the redundant data representation in which a motion sequence is 263 dimensional and includes `<root-rotation, root-position, joint-positions-wrt-root, joint-rotations-wrt-root, joint-velocities, foot-contacts>`. This is sample is then processed resulting in a 22x3 dimensional vector that represents the 3D joint positions for all 22 joints. The processed output is then visualized using the visualization part of the code. 

Best,
Setareh

**Comment by Stefano-retinize** [comment]
I think what @XH-B means is that actually that root position within the R^263 vector isn't the global root position. It seems is the local root position respect to another object. Indeed to built the joint position this root position is added in the recover_from_ric method:

````
def recover_from_ric(data, joints_num, abs_3d=False):
    r_rot_quat, r_pos = recover_root_rot_pos(data, abs_3d=abs_3d) #root quat, and root_pos, global?
    positions = data[..., 4:(joints_num - 1) * 3 + 4]  #selecting features from 4 to 67. Local pos respect root.
    positions = positions.view(positions.shape[:-1] + (-1, 3))

    '''Add Y-axis rotation to local joints'''
    positions = qrot(qinv(r_rot_quat[..., None, :]).expand(positions.shape[:-1] + (4,)), positions)

    '''Add root XZ to joints'''
    -->positions[..., 0] += r_pos[..., 0:1]
    -->positions[..., 2] += r_pos[..., 2:3]

    '''Concate root and joints'''
    positions = torch.cat([r_pos.unsqueeze(-2), positions], dim=-2)

    return positions
````

but those values aren't global and this can be seen in the .mp4 plotted videos where the video walks on the side. I can't figure it out why is that yet. Could it be because it has as input the root translation? if so, where in the code is?


https://github.com/user-attachments/assets/bcd5f3f7-909b-4f78-92ca-5931dd0fa792

**Comment by setarehc** [comment]
Are you using the pretrained models? If yes, they are trained with the absolute representations (in which the root positions are global instead of relative) and the samples taken from the model should have that representation as well. If you've trained your own model or if you are doing conditional sampling or editing, make sure your downloaded data has the absolute-root representations by following the instructions under `Full data` aquisition for GMD in the README. Lastly, if you use the absolute-root representations for the data, to process the output sample, `recover_from_ric` must be called with `abs_3d` set to `True`. I think that should be the case by default in the code, but do double-check this. 

-Setareh

**Comment by Stefano-retinize** [comment]
I indeed was using the vanilla motion_representation.ipynb and now I can see translation. But the animation doesn't look good. Am I probably doing something wrong, right? this isn't the model performance:


https://github.com/user-attachments/assets/2c7e1637-6837-4af8-a40e-b139d4581bbe

Thanks!

**Comment by setarehc** [comment]
This output is not model's normal performance. There seems to be a mismatch between the data and the model's root positions visible by the little jumps over keyframes. How do your dataset samples look if you visualize them?

**Comment by StefanoBraghetto** [comment]
I am trying to animate it using local rotation with wrt parents, for that I'm trying:
```
local_rot = sample[..., 67:(joints_num - 1) * 6 + 67].reshape(1, 196, joints_num-1, -1) #Sample is the output after inv_transform
r_rot_quat, r_pos = recover_root_rot_pos(sample, abs_3d=True)
local_rot = cont6d_to_matrix(local_rot)
global_rot = torch.matmul(local_rot  , quaternion_to_matrix(r_rot_quat))
```

#to obtain respect parent
```
inv_parent_global_rot = torch.inverse(global_rot)
for joint_idx:
     local_rot_wrt_parent[:, :, joint_idx - 1] = torch.matmul(inv_parent_global_rot[parent[joint_idx]], global_rot[:, :, joint_idx])
```


I guess my question can be simplified to: If I transform the features from 67 to 193 in the output to a rotation matrix: `local_rot_mat`  and I do:
`local_rot_mat * global_root_rot`, will I be obtaining the global rotation of each joint?


Also, are the joints index the same as the SMPL?
![image](https://github.com/user-attachments/assets/e4dde97a-0a21-46d3-8e6f-c36c77bcad29)

**Comment by Stefano-retinize** [comment]
I've been trying to solve the last question. And I think I reached to a conclusion. If it's possible could you please cnfirm that the conclusion is right?
The main question above was:

```
I guess my question can be simplified to: If I transform the features from 67 to 193 in the output to a rotation matrix: local_rot_mat and I do:
local_rot_mat * global_root_rot, will I be obtaining the global rotation of each joint?
```

And I think the problem is the global_root_rot. Since the root rotation is only represented by the Y axis rotation. It's not possible to obtain from the AI model prediction the root rotation. So in other words, the features from 67 to 193 are rotation wrt the root rot but the model is not giving the root rot.
Thanks for your help!

**Comment by setarehc** [comment]
What you describe could be the reason for the mismatch.
Closing for now, please feel free to reopen.
